### DISCUSSION TITLE
Variable Explorer contribution

### DISCUSSION THREAD

**Discussion by miduncan** [nteract/archived-desktop-app#5584] [Ideas]
Hey all!

Currently my team has a feature called “Variable Explorer”, which is a table that shows the current values of the variables in the kernel. I’ve attached a gif at the bottom to see it in action.

We’ve written this so that we can contribute the logic back to nteract (perhaps in its own package?). Essentially what would be added is 

-	Redux State
o	Array of variables (name, type, length, value)
o	isLoading (used for loading indicators when querying kernel)
o	isEnabled (so it can be disabled for performance reasons when not shown)
o	Provider used for querying kernel (only supports ipython currently)
-	Actions
o	Set for redux state fields above
o	RefreshVariables to re-query kernel
-	Selectors
o	Get for redux state fields above
-	Epics
o	After a RUN_CELL, enableVariableExplorer, or refreshVariables action, the variable values will be refreshed and stored variables array in the redux state

Unfortunately, we don’t have the time allocated for creating a UI for this; however, since the variable info would all be accessible via selectors, this should be easy to hook-up for future UI efforts. Would this be something that nteract is interested in? 

Thanks,

Michael

![image](https://aznb.azurewebsites.net/assets/images/variable-explorer-2dc79f29debca215c20c75bdc0045556.gif)

**Comment by miduncan** [comment]
@captainsafia Thoughts?

**Comment by captainsafia** [comment]
> We’ve written this so that we can contribute the logic back to nteract (perhaps in its own package?). Essentially what would be added is

This looks like a sensible breakdown for the API.

> Unfortunately, we don’t have the time allocated for creating a UI for this;

I have some hesitancies about this. In the past, we've gotten in trouble with implementation just the API for a feature without implementing the UI. For one, it means that we don't have a way to verify the behavior of something end-to-end and it inevitably brings up requests for the UI at some point.

I certainly don't want my perspective to be the hold up here but I think it's one of those things where it is best to go all the way with it.

@miduncan Based on your experience, do you think building out the UI is a large investment or is most of the hard work in figuring out the APIs and integration with the kernel?

**Reply by miduncan** [reply]
> do you think building out the UI is a large investment or is most of the hard work in figuring out the APIs and integration with the kernel?

No, the UI is a fairly small investment. Once the epics are running, you just need to use the enable action (so it is actively querying the kernel) and then the variables are all in the state for you. So basically, the UI could be as simple as reading the state and putting rows in a table and then the rest is just styling. For our implementation, it took me less than a day.

The "gotcha" I would worry about for nteract is that I'm not sure if there is a concept of a "panel" where we could put this UI. If not, there is some thinking to be done design-wise for what is supported, where it will be, etc.. Is there currently anything that solves this problem?

**Reply by captainsafia** [reply]
So, at least for the desktop app, we can see if there is a straightforward way for us to show this data in the Chrome DevTools panel although that has the consequence of not being super discoverable. We can also see how easy it is to to reuse the table of contents panel on the side as a bottom panel.

Let's give it two days UI cost given the estimate you gave + any additional integration work for the UI.

I'm fine integrating just the code in that case. We can try to get another NumFOCUS grant to complete the UI work for this (and all the other non-UI filled stubs we have).

Thanks for the follow-up!
